File: ddns/provider/namesilo.py
# coding=utf-8
"""
NameSilo API Provider
DNS provider implementation for NameSilo domain registrar
@doc: https://www.namesilo.com/api-reference
@author: NewFuture & Copilot
"""

from ._base import BaseProvider, TYPE_JSON


class NamesiloProvider(BaseProvider):
    """
    NameSilo DNS API Provider

    Supports DNS record management through NameSilo's API including:
    - Domain information retrieval
    - DNS record listing
    - DNS record creation
    - DNS record updating
    """

    endpoint = "https://www.namesilo.com"
    content_type = TYPE_JSON

    def _validate(self):
        """Validate authentication credentials"""
        # NameSilo only requires API key (token), not ID
        if not self.token:
            raise ValueError("API key (token) must be configured for NameSilo")
        if not self.endpoint:
            raise ValueError("API endpoint must be defined in {}".format(self.__class__.__name__))

        # Warn if ID is configured since NameSilo doesn't need it
        if self.id:
            self.logger.warning("NameSilo does not require 'id' configuration - only API key (token) is needed")

        # Show pending verification warning
        self.logger.warning("NameSilo provider implementation is pending verification - please test thoroughly")

    def _request(self, operation, **params):
        # type: (str, **(str | int | bytes | bool | None)) -> dict|None
        """
        Send request to NameSilo API

        Args:
            operation (str): API operation name
            params: API parameters

        Returns:
            dict: API response data
        """
        # Filter out None parameters
        params = {k: v for k, v in params.items() if v is not None}

        # Add required authentication and format parameters
        params.update({"version": "1", "type": "json", "key": self.token})

        # Make API request
        response = self._http("GET", "/api/" + operation, queries=params)

        # Parse response
        if response and isinstance(response, dict):
            reply = response.get("reply", {})

            # Check for successful response
            if reply.get("code") == "300":  # NameSilo success code
                return reply
            else:
                # Log error details
                error_msg = reply.get("detail", "Unknown error")
                self.logger.warning("NameSilo API error [%s]: %s", reply.get("code", "unknown"), error_msg)

        return None

    def _query_zone_id(self, domain):
        # type: (str) -> str | None
        """
        Query domain information to get domain as zone identifier
        @doc: https://www.namesilo.com/api-reference#domains/get-domain-info
        """
        response = self._request("getDomainInfo", domain=domain)

        if response:
            # Domain exists, return the domain name as zone_id
            domain_info = response.get("domain", {})
            if domain_info:
                self.logger.debug("Domain found: %s", domain)
                return domain

        self.logger.warning("Domain not found or not accessible: %s", domain)
        return None

    def _query_record(self, zone_id, subdomain, main_domain, record_type, line, extra):
        # type: (str, str, str, str, str | None, dict) -> dict | None
        """
        Query existing DNS record
        @doc: https://www.namesilo.com/api-reference#dns/list-dns-records
        """
        response = self._request("dnsListRecords", domain=main_domain)

        if response:
            records = response.get("resource_record", [])

            # Find matching record
            for record in records:
                if record.get("host") == subdomain and record.get("type") == record_type:
                    self.logger.debug("Found existing record: %s", record)
                    return record

        self.logger.debug("No matching record found for %s.%s (%s)", subdomain, main_domain, record_type)
        return None

    def _create_record(self, zone_id, subdomain, main_domain, value, record_type, ttl, line, extra):
        # type: (str, str, str, str, str, int | str | None, str | None, dict) -> bool
        """
        Create new DNS record
        @doc: https://www.namesilo.com/api-reference#dns/add-dns-record
        """
        response = self._request(
            "dnsAddRecord", domain=main_domain, rrtype=record_type, rrhost=subdomain, rrvalue=value, rrttl=ttl
        )

        if response:
            record_id = response.get("record_id")
            self.logger.info("DNS record created successfully: %s", record_id)
            return True
        else:
            self.logger.error("Failed to create DNS record")
            return False

    def _update_record(self, zone_id, old_record, value, record_type, ttl, line, extra):
        # type: (str, dict, str, str, int | str | None, str | None, dict) -> bool
        """
        Update existing DNS record
        @doc: https://www.namesilo.com/api-reference#dns/update-dns-record
        """
        record_id = old_record.get("record_id")
        if not record_id:
            self.logger.error("No record_id found in old_record: %s", old_record)
            return False

        # In NameSilo, zone_id is the main domain name
        response = self._request(
            "dnsUpdateRecord",
            rrid=record_id,
            domain=zone_id,  # zone_id is main_domain in NameSilo
            rrhost=old_record.get("host"),  # host field contains subdomain
            rrvalue=value,
            rrtype=record_type,
            rrttl=ttl or old_record.get("ttl"),
        )

        if response:
            self.logger.info("DNS record updated successfully: %s", record_id)
            return True
        else:
            self.logger.error("Failed to update DNS record")
            return False

File: rust/src/provider/namesilo.rs
use std::collections::BTreeMap;

use serde_json::Value;

use crate::error::{Error, Result};
use crate::http::Method;

use super::base::{CrudProvider, ProviderContext, RecordRequest, value_to_string};

pub struct NamesiloProvider<'a> {
    context: ProviderContext<'a>,
    zones: BTreeMap<String, String>,
}

impl<'a> NamesiloProvider<'a> {
    pub fn new(context: ProviderContext<'a>) -> Result<Self> {
        if context.token.is_empty() {
            return Err(Error::Config(
                "NameSilo API key must be configured".to_owned(),
            ));
        }
        Ok(Self {
            context,
            zones: BTreeMap::new(),
        })
    }

    fn api(&self, action: &str, mut parameters: BTreeMap<String, String>) -> Result<Option<Value>> {
        parameters.extend([
            ("version".to_owned(), "1".to_owned()),
            ("type".to_owned(), "json".to_owned()),
            ("key".to_owned(), self.context.token.to_owned()),
        ]);
        let response = self.context.send_json(
            Method::Get,
            &format!("/api/{action}"),
            &parameters,
            None,
            BTreeMap::new(),
        )?;
        let reply = response.get("reply").cloned().unwrap_or(Value::Null);
        let code = reply
            .get("code")
            .and_then(Value::as_str)
            .unwrap_or("unknown");
        match code {
            "300" => Ok(Some(reply)),
            // Code 200 means the domain is not active or not in this account.
            "200" if action == "getDomainInfo" => Ok(None),
            _ => {
                let detail = reply
                    .get("detail")
                    .and_then(Value::as_str)
                    .unwrap_or("unknown error");
                Err(Error::Provider(
                    self.context
                        .logger
                        .mask(&format!("NameSilo API error {code}: {detail}")),
                ))
            }
        }
    }
}

impl CrudProvider for NamesiloProvider<'_> {
    fn context(&self) -> &ProviderContext<'_> {
        &self.context
    }
    fn zone_cache(&mut self) -> &mut BTreeMap<String, String> {
        &mut self.zones
    }
    fn query_zone_id(&mut self, domain: &str) -> Result<Option<String>> {
        self.api(
            "getDomainInfo",
            BTreeMap::from([("domain".to_owned(), domain.to_owned())]),
        )?
        .map(|reply| {
            reply
                .get("domain")
                .filter(|value| !value.is_null())
                .map(|_| domain.to_owned())
                .ok_or_else(|| {
                    Error::Provider("NameSilo returned invalid domain information".to_owned())
                })
        })
        .transpose()
    }
    fn query_record(
        &mut self,
        _zone_id: &str,
        subdomain: &str,
        main_domain: &str,
        request: &RecordRequest<'_>,
    ) -> Result<Option<Value>> {
        let reply = self
            .api(
                "dnsListRecords",
                BTreeMap::from([("domain".to_owned(), main_domain.to_owned())]),
            )?
            .ok_or_else(|| Error::Provider("NameSilo returned no record list".to_owned()))?;
        let records = reply
            .get("resource_record")
            .and_then(Value::as_array)
            .ok_or_else(|| {
                Error::Provider("NameSilo returned an invalid record list".to_owned())
            })?;
        for record in records {
            let record_host = record
                .get("host")
                .and_then(Value::as_str)
                .ok_or_else(|| Error::Provider("NameSilo record has no host".to_owned()))?;
            let record_type = record
                .get("type")
                .and_then(Value::as_str)
                .filter(|value| !value.is_empty())
                .ok_or_else(|| Error::Provider("NameSilo record has no type".to_owned()))?;
            if record_host == subdomain && record_type == request.record_type {
                return Ok(Some(record.clone()));
            }
        }
        Ok(None)
    }
    fn create_record(
        &mut self,
        _zone_id: &str,
        subdomain: &str,
        main_domain: &str,
        request: &RecordRequest<'_>,
    ) -> Result<()> {
        let mut parameters = BTreeMap::from([
            ("domain".to_owned(), main_domain.to_owned()),
            ("rrtype".to_owned(), request.record_type.to_owned()),
            ("rrhost".to_owned(), subdomain.to_owned()),
            ("rrvalue".to_owned(), request.address.to_owned()),
        ]);
        if let Some(ttl) = request.ttl {
            parameters.insert("rrttl".to_owned(), ttl.to_string());
        }
        let reply = self.api("dnsAddRecord", parameters)?;
        if reply
            .as_ref()
            .and_then(|reply| reply.get("record_id"))
            .is_some()
        {
            Ok(())
        } else {
            Err(Error::Provider(
                "NameSilo failed to create record".to_owned(),
            ))
        }
    }
    fn update_record(
        &mut self,
        zone_id: &str,
        record: &Value,
        request: &RecordRequest<'_>,
    ) -> Result<()> {
        let record_id = record
            .get("record_id")
            .and_then(value_to_string)
            .ok_or_else(|| Error::Provider("NameSilo record has no record_id".to_owned()))?;
        let mut parameters = BTreeMap::from([
            ("rrid".to_owned(), record_id),
            ("domain".to_owned(), zone_id.to_owned()),
            (
                "rrhost".to_owned(),
                record
                    .get("host")
                    .and_then(value_to_string)
                    .unwrap_or_default(),
            ),
            ("rrvalue".to_owned(), request.address.to_owned()),
            ("rrtype".to_owned(), request.record_type.to_owned()),
        ]);
        if let Some(ttl) = request.ttl.or_else(|| {
            record
                .get("ttl")
                .and_then(value_to_string)
                .and_then(|ttl| ttl.parse::<u32>().ok())
        }) {
            parameters.insert("rrttl".to_owned(), ttl.to_string());
        }
        self.api("dnsUpdateRecord", parameters)?
            .map(|_| ())
            .ok_or_else(|| Error::Provider("NameSilo failed to update record".to_owned()))
    }
}
